### PR TITLE
Fix register lowering from FIRRTL

### PR DESCRIPTION
**Pull request by darthscsi**
Handle registers with optional synchronous reset.
Fixes issue of reset happening in initial block only (reset != initialization).
Rename reginit to regreset since the semantics of reset, which are modeled, say nothing about initialization.

### CODE REVIEW

**Comment by darthscsi** [comment]
A fair question is whether we need a register with synchronous reset op, since that can be modeled as a register + mux.  We probably do need an asynchronous reset register (in firrtl, we can tell by type, but thinking ahead to RTL...).

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp [hunk 1]
```diff
@@ -1131,26 +1191,29 @@ LogicalResult FIRRTLLowering::visitDecl(RegOp op) {
...
+
   if (!resultType || !resetSignal || !resetValue)
     return failure();
 
   auto regResult = builder->create<sv::RegOp>(resultType, op.nameAttr());
   setLowering(op, regResult);
 
+  inBlocksForReg(
+      clockVal, resetSignal,
+      [&]() { builder->create<sv::BPAssignOp>(regResult, resetValue); },
```
Comment (seldridge) [inline lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:1208]: Use a non-blocking statement here as I think that's right for an always block and it's also consistent with the rest of the logic and what the SFC does here.

```suggestion
      [&]() { builder->create<sv::PAssignOp>(regResult, resetValue); },
```

File: include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td [hunk 2]
```diff
@@ -124,12 +124,12 @@ def RegOp : FIRRTLOp<"reg", [/*MemAlloc*/]> {
   }];
 }
 
-def RegInitOp : FIRRTLOp<"reginit", [/*MemAlloc*/]> {
+def RegResetOp : FIRRTLOp<"regreset", [/*MemAlloc*/]> {
   let summary = "Define a new register with an initializer";
```
Comment (lattner) [inline include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td:128]: Plz update the "with an initializer" in the summary

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp [hunk 3]
```diff
@@ -863,6 +866,14 @@ struct FIRRTLLowering : public LowerFIRRTLToRTLBase<FIRRTLLowering>,
   /// key should have a FIRRTL type, the result will have an RTL dialect type.
   DenseMap<Value, Value> valueMapping;
 
+  /// Each register lowered needs to have connects placed in the always block,
+  /// and the else reset block.  This tracks the block for each lowered
+  /// register.
+  DenseMap<std::pair<Value, Value>, std::pair<Block *, Block *>> regBlocks;
+
+  /// Each value used as a clock generates one always block.
+  DenseMap<Value, Block *> clockBlocks;
```
Comment (lattner) [inline lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:875]: The general structure of LowerToRTL is to emit new always/if/initial etc blocks, and have a cleanup pass (yet to be written) that merges them together, folding conditions etc.

Does this structure work this this, and eliminate the need for the maps?  It looks like these exist just to make prettier IR.

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp [hunk 4]
```diff
@@ -1025,6 +1036,50 @@ LogicalResult FIRRTLLowering::setLoweringTo(Operation *orig,
   return setLowering(orig->getResult(0), result);
 }
 
+/// Create, if needed an always block for the given clock value
+void FIRRTLLowering::inAlwaysForClock(Value clock, std::function<void()> f) {
```
Comment (lattner) [inline lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:1040]: I'm not sure if these functions will be necessary after the above comment, but if so, please give them a verb'y name, like "emitAlwaysForClock"

**Review by lattner — COMMENTED** [review]
Oh good catch, thank you for tackling this!

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp [hunk 5]
```diff
@@ -1025,6 +1036,50 @@ LogicalResult FIRRTLLowering::setLoweringTo(Operation *orig,
...
+  auto block = clockBlocks[clock];
+  auto abuilder =
+      ImplicitLocOpBuilder::atBlockTerminator(builder->getLoc(), block);
+  auto oldBuilder = builder;
+  builder = &abuilder;
+  f();
+  builder = oldBuilder;
+}
+
+void FIRRTLLowering::inBlocksForReg(Value clock, Value reset,
```
Comment (seldridge) [inline lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:1054]: Documentation comment would be good here.

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp [hunk 6]
```diff
@@ -1025,6 +1036,50 @@ LogicalResult FIRRTLLowering::setLoweringTo(Operation *orig,
...
 
+/// Create, if needed an always block for the given clock value
+void FIRRTLLowering::inAlwaysForClock(Value clock, std::function<void()> f) {
+  if (!clockBlocks.count(clock))
+    clockBlocks[clock] =
+        builder->create<sv::AlwaysOp>(EventControl::AtPosEdge, clock)
+            .getBodyBlock();
+  auto block = clockBlocks[clock];
+  auto abuilder =
+      ImplicitLocOpBuilder::atBlockTerminator(builder->getLoc(), block);
```
Comment (seldridge) [inline lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:1047]: `block` only has the one use so this could be inlined.

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp [hunk 7]
```diff
@@ -1025,6 +1036,50 @@ LogicalResult FIRRTLLowering::setLoweringTo(Operation *orig,
...
+  auto blocks = regBlocks[std::make_pair(clock, reset)];
+  auto oldBuilder = builder;
+
+  auto abuilder =
+      ImplicitLocOpBuilder::atBlockTerminator(builder->getLoc(), blocks.first);
+  builder = &abuilder;
+  freset();
+
+  auto bbuilder =
+      ImplicitLocOpBuilder::atBlockTerminator(builder->getLoc(), blocks.second);
```
Comment (seldridge) [inline lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:1076]: The PR could re-assign to `abuilder` and avoid defining `bbuilder` here.

File: include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td [hunk 8]
```diff
@@ -124,12 +124,12 @@ def RegOp : FIRRTLOp<"reg", [/*MemAlloc*/]> {
...
   let description = [{
     Declare a new register:
     ```
-      %name = firrtl.reginit %clockVal, %resetSignal, %resetValue : t1
+      %name = firrtl.regreset %clockVal, %resetSignal, %resetValue : t1
     ```
     }];
```
Comment (seldridge) [inline include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td:134]: The name change feels somewhat unnecessary relative to the functional content of the PR. I'm fine to keep this if you think it's more clear this way.

**Review by seldridge — APPROVED** [review]
:100: 

Some (nitty) comments throughout. Content looks great as does the output Verilog. Not sure how we missed this. 😬

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp (hunk 1 above)
Comment (seldridge) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:1208]: :point_up: This is the only change that I would definitely like in there.

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp (hunk 3 above)
Comment (darthscsi) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:875]: So it turns out no.  Registers of different type (asyncreset, syncreset, noreset) need to be in separate always blocks.
http://www.sunburst-design.com/papers/CummingsSNUG2003Boston_Resets.pdf

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp (hunk 4 above)
Comment (darthscsi) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:1040]: Dropping the always merging for now, given complication in previous comment.

File: lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp (hunk 3 above)
Comment (darthscsi) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:875]: What isn't clear is whether tools would be ok with this when mixing followers and reset registers:
```
always @posedge clk begin
  followerreg <=  val
  if (reset) begin
     syncreg <= val2
   end else begin
     syncreg <= val3
  end
end
```

Comment (seldridge) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:875]: Probably not. There may also be additional restrictions around asynchronous resets like [the Formality requirement of a single statement in an asynchronous reset always block](https://github.com/chipsalliance/firrtl/blob/15013df6f6ac2dafeb35d7ed15cf95c7ac8a5bef/src/main/scala/firrtl/backends/verilog/VerilogEmitter.scala#L573).

I'd be in favor of one block for each `(Clock, Option<ResetType>)` tuple and not merge beyond that.

Comment (lattner) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:875]: Ok, I wasn't aware of that.  What does that mean for the semantics of always?  How do we reason about what can and what can't be merged?

Comment (seldridge) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp:875]: Merging after we're at the level of `always` may be tricky as we've lost type information about what is a reset. E.g., it's unclear in the above example that `reset` is actually a synchronous reset and should block the merge.

If this is coming from FIRRTL, the information is available and merging can happen, but this conflicts with the approach of lowering to `always` and then merging. Maybe this motivates another IR construct that represents registers or a "higher level" always block that takes an `(Clock, Option<ResetType>)` input? (This pushes towards an IR construct for registers like the "sequential" dialect discussion.)

Comment (darthscsi) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp]: I think much of this goes away when the RTL dialect has a register abstraction...

Some of the complication in merging will come from having reset and !reset floating around as conditions.

But for now, I've updated to go back to the always blocks per statement.

It seems we may have to tread always blocks as having a list of sensitvities and a reset signal so we can keep things sorted out.

Comment (lattner) [reply lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp]: > It seems we may have to tread always blocks as having a list of sensitvities and a reset signal so we can keep things sorted out.

Yeah I think this is the right path.  We can have two forms of always blocks:  

1) the current syntactic one which is basically a "don't touch" for the optimizer because the frontend didn't provide metadata that makes it safe to do various transformations (I still don't grok exactly why, is this pattern matching on the part of downstream tools? ELI5 plz :-)

2) the current one with a list of additional metadata (that the frontend can optionally provide) which, when present, allows safe merging and transformation.  FIRRTL lowering can provide this if/since we already have it.

I think that this is going to be non-obvious, so splitting it out into its own discussion (probably on the forum) would make sense.  In the meantime, thank you for doing the "technically wrong but consistent with the current direction" approach @darthscsi

**Review by lattner — APPROVED** [review]
Patch looks really great, thanks!
